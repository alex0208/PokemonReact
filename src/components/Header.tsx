import {
    AppBar,
    Toolbar,
    Typography,
    IconButton,
    styled,
    alpha,
    InputBase,
    DialogTitle,
    Dialog,
    Button,
    DialogActions,
    DialogContent,
    DialogContentText,
    TextField,
    FormControl,
    InputLabel,
    MenuItem,
    Select,
    SelectChangeEvent,
} from '@mui/material';
import FilterListIcon from '@mui/icons-material/FilterList';
import SearchIcon from '@mui/icons-material/Search';
import { ChangeEvent, Dispatch, SetStateAction, useState } from 'react';
import { SORTING_DIRECTION } from '../common/constants/sort';
import { order_by } from '../graphql/generated/globalTypes';

interface Props {
    setSearchTerm: Dispatch<SetStateAction<string>>;
    // searchingDirection: any
}

// const Header = ({ setSearchTerm, sortingDirection }: Props) => {
const Header = ({
    setSearchTerm,
    orderBy: [orderBy, setOrderBy],
    sortingDirection: [sortingDirection, setSortingDirection],
    pageSize,
    setPageSize,
}: any) => {
    const handleChange = (
        e: ChangeEvent<HTMLTextAreaElement | HTMLInputElement>
    ) => {
        e.preventDefault();
        setSearchTerm(e.target.value);
    };

    const [open, setOpen] = useState(false);

    const handleClickOpen = () => {
        setOpen(true);
    };

    const handleClose = (event = {}, reason = '') => {
        setOpen(false);
    };

    return (
        <AppBar position="static" sx={{ mb: 8 }}>
            <Toolbar>
                <Typography variant="h6" component="div" sx={{ flexGrow: 1 }}>
                    Pokemon
                </Typography>
                <Search>
                    <SearchIconWrapper>
                        <SearchIcon />
                    </SearchIconWrapper>
                    <StyledInputBase
                        placeholder="Searchâ€¦"
                        inputProps={{ 'aria-label': 'search' }}
                        onChange={handleChange}
                    />
                </Search>
                <IconButton
                    onClick={handleClickOpen}
                    size="large"
                    edge="start"
                    color="inherit"
                    aria-label="menu"
                    sx={{ mr: 2 }}
                >
                    <FilterListIcon />
                </IconButton>

                <Dialog
                    open={open}
                    onClose={(event, reason) => {
                        if (reason !== 'backdropClick') {
                            handleClose(event, reason);
                        }
                    }}
                >
                    <DialogTitle>Search settings</DialogTitle>
                    <DialogContent sx={{ minWidth: 300 }}>
                        <FormControl fullWidth sx={{ mt: 3 }}>
                            <InputLabel id="demo-simple-select-label">
                                Sort by
                            </InputLabel>
                            <Select
                                labelId="demo-simple-select-label"
                                id="demo-simple-select"
                                value={orderBy}
                                label="Sort by"
                                onChange={(e: SelectChangeEvent<string>) =>
                                    setOrderBy(e.target.value)
                                }
                            >
                                <MenuItem value={SORTING_DIRECTION.NAME}>
                                    Name
                                </MenuItem>
                                <MenuItem value={SORTING_DIRECTION.HEIGHT}>
                                    Height
                                </MenuItem>
                                <MenuItem value={SORTING_DIRECTION.WEIGHT}>
                                    Weight
                                </MenuItem>
                            </Select>
                        </FormControl>

                        <FormControl fullWidth sx={{ mt: 3 }}>
                            <InputLabel id="demo-simple-select-label">
                                Sorting direction
                            </InputLabel>
                            <Select
                                labelId="demo-simple-select-label"
                                id="demo-simple-select"
                                value={sortingDirection}
                                label="Sort direction"
                                onChange={(e: SelectChangeEvent<string>) =>
                                    setSortingDirection(e.target.value)
                                }
                            >
                                <MenuItem value={order_by.asc}>
                                    Ascending
                                </MenuItem>
                                <MenuItem value={order_by.desc}>
                                    Descending
                                </MenuItem>
                            </Select>
                        </FormControl>
                        <FormControl fullWidth sx={{ mt: 3 }}>
                            <InputLabel id="demo-simple-select-label">
                                Page size
                            </InputLabel>
                            <Select
                                labelId="demo-simple-select-label"
                                id="demo-simple-select"
                                value={pageSize}
                                label="Sort direction"
                                onChange={(e: SelectChangeEvent<number>) =>
                                    setPageSize(e.target.value)
                                }
                            >
                                <MenuItem value={10}>10</MenuItem>
                                <MenuItem value={20}>20</MenuItem>
                                <MenuItem value={50}>50</MenuItem>
                            </Select>
                        </FormControl>
                    </DialogContent>
                    <DialogActions>
                        <Button onClick={handleClose}>Cancel</Button>
                        <Button onClick={handleClose}>Save</Button>
                    </DialogActions>
                </Dialog>
            </Toolbar>
        </AppBar>
    );
};

const Search = styled('div')(({ theme }) => ({
    position: 'relative',
    borderRadius: theme.shape.borderRadius,
    backgroundColor: alpha(theme.palette.common.white, 0.15),
    '&:hover': {
        backgroundColor: alpha(theme.palette.common.white, 0.25),
    },
    marginLeft: 0,
    marginRight: 20,
    width: '100%',
    [theme.breakpoints.up('sm')]: {
        marginLeft: theme.spacing(1),
        width: 'auto',
    },
}));

const SearchIconWrapper = styled('div')(({ theme }) => ({
    padding: theme.spacing(0, 2),
    height: '100%',
    position: 'absolute',
    pointerEvents: 'none',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
}));

const StyledInputBase = styled(InputBase)(({ theme }) => ({
    color: 'inherit',
    '& .MuiInputBase-input': {
        padding: theme.spacing(1, 1, 1, 0),
        // vertical padding + font size from searchIcon
        paddingLeft: `calc(1em + ${theme.spacing(4)})`,
        transition: theme.transitions.create('width'),
        width: '100%',
        [theme.breakpoints.up('sm')]: {
            width: '12ch',
            '&:focus': {
                width: '20ch',
            },
        },
    },
}));

export default Header;
