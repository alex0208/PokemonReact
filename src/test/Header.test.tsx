import { MemoryRouter } from 'react-router';
import renderer from 'react-test-renderer';
import Header from '../components/Header';

it('matches snapshot', () => {
    const data = {
        searchTerm: [],
        orderBy: [],
        sortingDirection: [],
        pageSize: 20,
        setPageSize: () => {},
    };

    const tree = renderer
        .create(
            <MemoryRouter>
                <Header {...data} />
            </MemoryRouter>
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});
