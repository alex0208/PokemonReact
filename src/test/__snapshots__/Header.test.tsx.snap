// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<header
  className="MuiPaper-root-jqRKNJ dfuyOL MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root-gCJNDv kLzESQ MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic"
>
  <div
    className="MuiToolbar-root-hLnTOG SHPyl MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular"
  >
    <div
      className="MuiTypography-root-bWuLxc eoJEBL MuiTypography-root MuiTypography-h6"
    >
      Pokemon
    </div>
    <div
      className="sc-iqHYmW kqXCWp"
    >
      <div
        className="sc-crrszt jXrjiG"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root-ffGNMi ufROg MuiSvgIcon-root MuiSvgIcon-fontSizeMedium"
          data-testid="SearchIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
      </div>
      <div
        className="MuiInputBase-root-jytTJo iblFgs MuiInputBase-root MuiInputBase-colorPrimary sc-dQoVA dZtKAF"
        onClick={[Function]}
      >
        <input
          aria-label="search"
          className="MuiInputBase-input-hhOXZE iuLSHy MuiInputBase-input"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Searchâ€¦"
          type="text"
        />
      </div>
    </div>
    <button
      aria-label="menu"
      className="MuiButtonBase-root-JTLGk xjigV MuiButtonBase-root MuiIconButton-root-jYLena jOBJaf MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeLarge"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root-ffGNMi ufROg MuiSvgIcon-root MuiSvgIcon-fontSizeMedium"
        data-testid="FilterListIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M10 18h4v-2h-4v2zM3 6v2h18V6H3zm3 7h12v-2H6v2z"
        />
      </svg>
    </button>
  </div>
</header>
`;
